### ISSUE TITLE
Use for SentryPeer?

### ISSUE THREAD

**Issue by ghenry** [babelouest/huddersfield#2]
Hi Nicolas,

Could this be used for a 3rd party project? I'd like to use it for [SentryPeer](https://github.com/SentryPeer/SentryPeer) as well as [ulfius](https://github.com/babelouest/ulfius) and maybe [hoel](https://github.com/babelouest/hoel) if I ever do more than sqlite3 (which is amazing).

Thanks.

**Comment by babelouest** [comment]
Hello Gavin,

Of course you can use these scripts for another project, that's the beauty of OSS! ;)

I don't know if it's easy for another person to adapt it for its needs but I'll try to explain how it works:

The [Makefile](https://github.com/babelouest/huddersfield/blob/master/Makefile) is the entry point to run the build/test/smoke-test commands.
There is one target for each project/architecture/command, but to avoid more code duplication than it already has, there is a lot of federated code.
For example, to build ulfius on a debian stable architecture, the following commands are executed:

- get source of ulfius, yder and orcania and place the tar.gz in ulfius/
- build docker image debian/stable and name it `babelouest/deb`
- build docker image babelouest/ulfius based on `babelouest/deb` and name it `babelouest/ulfius`
- run docker image `babelouest/ulfius`, its < entrypoint.sh` script will compile and package ulfius and its dependencies, and copy the generated files in the host folder `ulfius/`

To build ulfius for debian testing or ubuntu lts, simply change the target architecture on the second step.

The `*-test` targets are used to run CI tests and make sure the test pass on all supported architectures

The `*.smoke` targets are used to install and run the program on the architecture to make sure the generated packages work fine, especially with the dependencies.

To add a new project in the build list, I would start by copying [hutch directory](https://github.com/babelouest/huddersfield/tree/master/hutch) and adapt the scripts to your project, example for a debian build:

- `$ cp -R hutch/ sentrypeer/`
- open `sentrypeer/build/deb/Dockerfile` and update the `apt-get install` packages with the ones you need
- open `sentrypeer/build/deb/entrypoint.sh` and update the build script to fit your needs

Then, in the [Makefile](https://github.com/babelouest/huddersfield/blob/master/Makefile) of the project, you must add a block for sentrypeer that looks like this:

```
sentrypeer-source: sentrypeer/sentrypeer.tar.gz

sentrypeer-deb:
	$(DOCKER_BUILD) -t babelouest/sentrypeer sentrypeer/build/deb/
	$(DOCKER_RUN) --rm -v $(shell pwd)/:/share babelouest/sentrypeer

sentrypeer-deb-test:
	cp sentrypeer/*.tar.gz sentrypeer/test/deb/
	$(DOCKER_BUILD) -t babelouest/sentrypeer-test sentrypeer/test/deb/
	rm -f sentrypeer/test/deb/*.tar.gz
	$(DOCKER_RUN) --rm -v $(shell pwd)/:/share babelouest/sentrypeer-test

sentrypeer-debian-stable: yder-source orcania-source sentrypeer-source
	$(MAKE) debian-stable
	$(MAKE) sentrypeer-deb
	xargs -a ./sentrypeer/packages -I% $(MAKE) upload-asset GITHUB_UPLOAD=$(GITHUB_UPLOAD) GITHUB_TOKEN=$(GITHUB_TOKEN) GITHUB_USER=$(GITHUB_USER) REPO=sentrypeer TAG=$(ULFIUS_VERSION) PATTERN=./sentrypeer/%

sentrypeer-debian-stable-test: sentrypeer-debian-stable
	$(MAKE) debian-stable
	$(MAKE) sentrypeer-deb-test
```

**Comment by ghenry** [comment]
Hi Nicolas,

Thank you very much for the detailed reply!

I'm looking to do an initial release soon and submit to Fedora, Ubuntu and Debian.

Have you automated or done that? I'll check your workflows. I received this advice from a very kind Debian dev:

> I recommend looking at https://wiki.debian.org/UpstreamGuide to some
> pointers about how to make it easy to ensure your software can be
> easily packaged in Debian.
> 
> If you're looking to create a deb package that is ready to be included
> in Debian it is best to have your debian/ directory on a separate
> branch. I recommend using git-buildpackage, see
> https://wiki.debian.org/PackagingWithGit (this is what most
> collaboratively maintained packages use now).
> 
> For information on how to create Debian packages see
> https://www.debian.org/doc/manuals/developers-reference/ (particularly
> chapters 5 and 6). If you want to look at more of the background there
> is the policy document, see https://www.debian.org/doc/debian-policy/
> 
> There are a lot of tools for efficiently building Debian packages. If
> you want to test your builds in a more controlled environment you
> should probably use pbuilder (of cowbuilder) because those use a clean
> chroot to build your packages to ensure you have the correct
> dependencies specified.
> 
> Hope this helps

**Comment by babelouest** [comment]
Hello again,

it's important to note that huddersfield is not intended to build packages that can be provided by the distributions, it's intended for me to provide packages compatible with the distribution I support.

I'm a Debian Developer myself and I package among other things my projects. But the packaging for Debian isn't based on this project, because the .deb generated by the cmake command `make package` is very different from the one generated by Debian's package builder toolset:
- the dependencies a somehow hardcoded in the cmake file
- there is no preinst or postinst scripts
- the debian package builder rebuilds everything, even the webapp when there is one
- etc.

The constraints are similar for other distributions, fedora requires its package builder toolset, arch as well, etc.

So basically, if you want to build packages for SentryPeer that you will master and will make them available in your [release pages](https://github.com/SentryPeer/SentryPeer/releases) or a download page somewhere, huddersfield can help you building packages and run tests easily.
But if you want to make official packages for distributions, you need to use each distribution official way to build packages, and so far I haven't seen any distribution that would accept cmake's `package` command (with reason).

**Comment by ghenry** [comment]
Thanks Nicolas. That all makes perfect sense.

What benefits do you have releasing your own packages? I presuming, testing, instant availability etc. Do you host any packaging infrastructure for apt / dnf etc?

Apologies for all the questions. It's just last night when I got your release message from github, I forgot I'd subscribed, as I found your project when I was looking for OAuth/OpenID systems and then spotted all your other projects which fit nicely with what I'm working on now :-D

Gavin.

**Comment by babelouest** [comment]
> What benefits do you have releasing your own packages? I presuming, testing, instant availability etc. Do you host any packaging infrastructure for apt / dnf etc?

Yes, the main advantage is to provide up-to-date packages for distributions that don't provide packages for my projects, or when a distrib releases a new version, like a non rolling distribution.
Also, thanks to huddersfield, I build and upload new packages as soon as I make a new release with a simple command: `make ulfius-build GITHUB_UPLOAD=1`, because of their ecosystems, distrib may spend more time to provide up to date packages.

> when I was looking for OAuth/OpenID systems and then spotted all your other projects which fit nicely with what I'm working on now :-D

You're welcome to reuse it, I'm glad that a member of the OpenLDAP team is interested with my work, especially since libldap is a non negligible component in my OSS work, see: https://github.com/babelouest/glewlwyd/blob/master/src/user/ldap.c

**Comment by ghenry** [comment]
Thanks Nicolas. That's what I thought. I was just pitching the idea for show on Packaging Open Source Software for [one of my shows](https://www.se-radio.net/team/gavin-henry/) for [Software Engineering Radio](https://www.se-radio.net/) so may reach out to you about it since you're an official Debian developer too?

That's a ton of code right there!!!!

**Comment by babelouest** [comment]
Sure, you can contact me to talk about packaging, but I think we should continue this conversation via e-mail then: mail@babelouest.org

**Comment by babelouest** [comment]
Hello @ghenry ,

Do you think this issue can be closed? Or do you have other questions about Huddersfield?

**Comment by ghenry** [comment]
Thanks Nicolas. Yep, closing now. I'll come back to the closed issue when I get to it. Again, thank you so much for your detailed replies and help.
